### PR TITLE
[Feature] ImVoteNet complete model

### PR DESCRIPTION
**Pull request by THU17cyz**
Added the ImVoteNet complete model, including initialization, multi-tower training and test (simple and aug).

### CODE REVIEW

File: mmdet3d/models/detectors/imvotenet.py [hunk 1]
```diff
@@ -1,51 +1,129 @@
...
+from mmdet.models import DETECTORS
+from .. import builder
 from .base import Base3DDetector
 
 
+class ImageMLPModule(nn.Module):
+
+    def __init__(self, input_dim=18, image_hidden_dim=256):
+        super().__init__()
+        self.img_feat_conv1 = nn.Conv1d(input_dim, image_hidden_dim, 1)
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/imvotenet.py:16]: May use conv_cfg and norm_cfg with build_conv_layer and build_norm_layer in case users want to try GN or other layers.

File: mmdet3d/models/detectors/imvotenet.py [hunk 2]
```diff
@@ -1,51 +1,129 @@
...
+        self.img_feat_conv2 = nn.Conv1d(image_hidden_dim, image_hidden_dim, 1)
+        self.img_feat_bn1 = nn.BatchNorm1d(image_hidden_dim)
+        self.img_feat_bn2 = nn.BatchNorm1d(image_hidden_dim)
+
+    def forward(self, img_features):
+        img_features = F.relu(
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/imvotenet.py:22]: May use inplace relu to save memory.

File: mmdet3d/models/detectors/imvotenet.py [hunk 3]
```diff
@@ -1,51 +1,129 @@
...
+            self.img_feat_bn2(self.img_feat_conv2(img_features)))
+
+        return img_features
+
+
+def sample_valid_seeds(mask, num_sampled_seed=1024):
+    device = mask.device
+    batch_size = mask.shape[0]
+    sample_inds = torch.zeros((batch_size, num_sampled_seed), device=device)
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/imvotenet.py:33]: mask.new_zeros?

File: mmdet3d/models/detectors/imvotenet.py [hunk 4]
```diff
@@ -1,51 +1,129 @@
...
+    for bidx in range(batch_size):
+        # return index of non zero elements
+        valid_inds = torch.arange(len(mask[bidx, :]))
+        if len(valid_inds) < num_sampled_seed:
+            rand_inds = np.random.choice(
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/imvotenet.py:38]: Can we use torch.random?

File: mmdet3d/models/detectors/imvotenet.py [hunk 5]
```diff
@@ -1,51 +1,129 @@
...
+
+
 @DETECTORS.register_module()
 class ImVoteNet(Base3DDetector):
     r"""`ImVoteNet <https://arxiv.org/abs/2001.10692>`_ for 3D detection."""
 
     def __init__(self,
                  pts_backbone=None,
-                 pts_bbox_head=None,
+                 pts_bbox_head_joint=None,
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/imvotenet.py:60]: can we make pts_bbox_head a list or a dict of config dicts?

File: mmdet3d/models/detectors/imvotenet.py [hunk 6]
```diff
@@ -1,51 +1,129 @@
...
+from torch.nn import functional as F
 
-from mmdet.models import DETECTORS, build_backbone, build_head, build_neck
+from mmdet3d.core import bbox3d2result, merge_aug_bboxes_3d
+from mmdet.models import DETECTORS
+from .. import builder
 from .base import Base3DDetector
 
 
+class ImageMLPModule(nn.Module):
```
Comment (ZwwWayne) [inline mmdet3d/models/detectors/imvotenet.py:12]: Move this module to ops.utils or models.utils

**Review by ZwwWayne — APPROVED** [review]

File: mmdet3d/models/detectors/imvotenet.py (hunk 5 above)
Comment (THU17cyz) [reply mmdet3d/models/detectors/imvotenet.py]: I plan to change it to a dict which consists of the common parts and the parts that belong to each head

File: mmdet3d/models/detectors/imvotenet.py [hunk 7]
```diff
@@ -1,51 +1,114 @@
...
 
-from mmdet.models import DETECTORS, build_backbone, build_head, build_neck
+from mmdet3d.core import bbox3d2result, merge_aug_bboxes_3d
+from mmdet3d.models.model_utils import ImageMLPModule
+from mmdet.models import DETECTORS
+from .. import builder
 from .base import Base3DDetector
 
 
+def sample_valid_seeds(mask, num_sampled_seed=1024):
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:12]: Docstring

File: mmdet3d/models/detectors/imvotenet.py [hunk 8]
```diff
@@ -1,51 +1,114 @@
...
+            rand_inds = torch.randperm(
+                len(valid_inds), device=device)[:num_sampled_seed]
+            cur_sample_inds = valid_inds[rand_inds]
+        sample_inds[bidx, :] = cur_sample_inds
+    return sample_inds.long()
+
+
 @DETECTORS.register_module()
 class ImVoteNet(Base3DDetector):
     r"""`ImVoteNet <https://arxiv.org/abs/2001.10692>`_ for 3D detection."""
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:42]: Docstring about input params for this class.

File: mmdet3d/models/model_utils/img_feat_mlp_module.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            Default: dict(type='Conv1d').
+        norm_cfg (dict): Config of normalization.
+            Default: dict(type='BN1d').
+        act_cfg (dict): Config of activation.
+            Default: dict(type='ReLU').
+    """
+
+    def __init__(self,
+                 in_channel=18,
+                 conv_channels=(256),
```
Comment (Tai-Wang) [inline mmdet3d/models/model_utils/img_feat_mlp_module.py:25]: (256, ) for tuple? Otherwise, the default input may trigger error (it will be regarded as int?).

**Review by Tai-Wang — COMMENTED** [review]
Just need some minor adjustments. There should be another two PRs about benchmark/configs and unit test I guess for complete the overall model.

**Comment by THU17cyz** [comment]
Just fixed some bugs after I tested the whole model. The next PR should be ready soon.

File: mmdet3d/models/detectors/imvotenet.py [hunk 10]
```diff
@@ -178,6 +353,12 @@ def forward_train(self,
...
                       gt_masks=None,
                       proposals=None,
+                      calib=None,
+                      bboxes_2d=None,
+                      gt_bboxes_3d=None,
+                      gt_labels_3d=None,
+                      pts_semantic_mask=None,
+                      pts_instance_mask=None,
                       **kwargs):
         """ Forward of training for image only or image and points.
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvotenet.py:363]: Add a blank line before Args.

**Review by Tai-Wang — APPROVED** [review]
Please finally check all the docstrings one more time and elaborate the incomplete ones. Also fix the failing of CI.

**Comment by THU17cyz** [comment]
For CI fix, refer to #350 @Tai-Wang
